File: Creational/AbstractFactory/Java/ThreeDShapeFactory.java
// ThreeDShapeFactory.java
public class ThreeDShapeFactory implements ShapeFactory {
  @Override
  public Shape createShape(ShapeType type) {
    switch (type) {
      case CIRCLE:
        return new Circle();
      case SQUARE:
        return new Square();
      default:
        throw new IllegalArgumentException("Unknown shape type: " + type);
    }
  }

  @Override
  public Renderer createRenderer() {
    return new ThreeDRenderer();
  }
}

File: Creational/AbstractFactory/JavaScript/threeDShapeFactory.js
// threeDShapeFactory.js

const ShapeFactory = require("./shapeFactory");
const ShapeType = require("./shapeType");
const Circle = require("./circle");
const Square = require("./square");
const ThreeDRenderer = require("./threeDRenderer");

class ThreeDShapeFactory extends ShapeFactory {
  createShape(shapeType) {
    switch (shapeType) {
      case ShapeType.CIRCLE:
        return new Circle();
      case ShapeType.SQUARE:
        return new Square();
      default:
        throw new Error(`Unknown shape type: ${shapeType}`);
    }
  }

  createRenderer() {
    return new ThreeDRenderer();
  }
}

module.exports = ThreeDShapeFactory;
